### PR TITLE
New interface traits

### PR DESCRIPTION
**Pull request by GrantM11235**

### CODE REVIEW

**Comment by GrantM11235** [comment]
Wow, I didn't expect this to impact performance much on my project (8 bit unbuffered parallel interface using `fill_contiguous`), but its over twice as fast as the master branch!

Please try it out, I think it might be faster than #146

**Review by almindor — CHANGES_REQUESTED** [review]
I generally like this approach it also simplifies the `Model` trait.

I think if you add CHANGELOG entry and migration documentation I'd prefer this over #146 

@rfuest what do you think?

**Review by almindor — APPROVED** [review]
LGTM! @rfuest I want to wait on your opinion as well however.

**Comment by rfuest** [comment]
I don't think it's a good idea to remove support for `display-interface` entirely at this point in time. We should at least have a fallback option for users that use custom `display-interface` impls, for example the FSMC support for STM32F4 (https://github.com/stm32-rs/stm32f4xx-hal/tree/master/src/fsmc_lcd).

**Comment by almindor** [comment]
> I don't think it's a good idea to remove support for `display-interface` entirely at this point in time. We should at least have a fallback option for users that use custom `display-interface` impls, for example the FSMC support for STM32F4 (https://github.com/stm32-rs/stm32f4xx-hal/tree/master/src/fsmc_lcd).

How would you want to address this though? Deprecate DI with 0.9 and give some time? Use features to switch between DI and "PI" versions? I don't want to overcomplitcate things or delay them unless necessary.

I think if we release this as 0.9, FSMC can keep supporting 0.8 until they reimplement for the new version. I find it interesting they decided to implement DI inside the HAL directly.

File: docs/MIGRATION.md
Comment (orsinium) [inline docs/MIGRATION.md:7]:
```suggestion
* The `display_interface` dependency has been removed in favor of our own traits and implementations. This gives significantly better performance (#149)
```

**Comment by GrantM11235** [comment]
I think it's technically possible to create an adapter to convert an old display_interface into the new trait:

<details>
<summary>expand me!</summary>

```rust
pub struct LegacyAdapter<DI: WriteOnlyDataCommand>(pub DI);

impl<DI: WriteOnlyDataCommand> Interface for LegacyAdapter<DI> {
    type PixelWord = u16;
    type Error = display_interface::DisplayError;

    fn send_command(&mut self, command: u8, args: &[u8]) -> Result<(), Self::Error> {
        self.0.send_commands(DataFormat::U8(&[command]))?;
        self.0.send_data(DataFormat::U8(args))
    }

    fn send_pixels<const N: usize>(
        &mut self,
        pixels: impl IntoIterator<Item = [Self::PixelWord; N]>,
    ) -> Result<(), Self::Error> {
        self.0
            .send_data(DataFormat::U16BEIter(&mut pixels.into_iter().flatten()))
    }

    fn send_repeated_pixel<const N: usize>(
        &mut self,
        pixel: [Self::PixelWord; N],
        count: u32,
    ) -> Result<(), Self::Error> {
        self.send_pixels((0..count).map(|_| pixel))
    }

    fn flush(&mut self) -> Result<(), Self::Error> {
        Ok(())
    }
}
```

</details>

I don't think it is a good idea for us to have this adapter though, because it doesn't work with Rgb666 and it doesn't solve any of the performance problems. You're better off just sticking with 0.8.0 instead.

I merged the `CommandInterface` and `PixelInterface` traits into a single `Interface` trait. The split was necessary in a previous version of this PR, but it isn't anymore.

almindor suggested putting the interface trait in it's own crate. I think this is probably the right thing to do eventually, but I would like to wait to see if the trait needs any more changes. It should be backwards compatible to move the trait to it's own crate as long as we publicly export it.

**Comment by almindor** [comment]
I think we should make a call. My take atm is that we should do this:
1. Separate the new trait into Write and Read version (read can be completely unimplemented at the moment). I was close to adding a DI read version ~1 year back but my MCU fluffed out on me
2. Separate the interface into its own crate, either in this repo or its own. @GrantM11235 I'll let you decide which way since its your code
3. Bump mipidsi version to 0.9 after and release

There's also the option of using the existing DI repo for this since I have write access, but I don't have admin so it'd be a cointoss on how that'd go longer term.

When it comes to [dependants of DI](https://crates.io/crates/display-interface/reverse_dependencies) I see only the two hal crates outside of direct drivers or implementations. I don't think it's worth doing some sort of deprecation delay. Looking at the code, it should be super trivial for someone to re-do the existing hal implementations for the new traits. And the nice thing is they can technically implement both alongside as well.

**Comment by GrantM11235** [comment]
There is nothing to split, this trait is already effectively the Write version. The Read version will almost certainly look something like `pub trait ReadInterface: Interface { /* ... */ }`, but that is outside the scope of this PR.

The main reason to put the trait in it's own crate is to provide external implementers with stability even when there are breaking changes to the main `mipidsi` crate. I think it is premature to do this now because we may or may not need to make changes to the trait. Once we are reasonably confident in the trait we can release `mipidsi-interface` (or whatever) as instant 1.0.0.

The trait is specifically made to be used by mipidsi, not any other types of displays, so I think it should stay in this repo even after it is put in it's own crate.

I agree that we should publish version 0.9 once this is merged :rocket:

**Comment by rfuest** [comment]
> I don't think it is a good idea for us to have this adapter though, because it doesn't work with Rgb666 and it doesn't solve any of the performance problems. You're better off just sticking with 0.8.0 instead.

We should at least mention this in the migration guide.

> There's also the option of using the existing DI repo for this since I have write access, but I don't have admin so it'd be a cointoss on how that'd go longer term.

We could also reuse the [`display-driver-hal`](https://github.com/embedded-graphics/display-driver-hal) repo, which was supposed to add higher level traits, like backlight control, but I don't see a reason not to also include the `display-interface` replacement.

> When it comes to [dependants of DI](https://crates.io/crates/display-interface/reverse_dependencies) I see only the two hal crates outside of direct drivers or implementations. I don't think it's worth doing some sort of deprecation delay. Looking at the code, it should be super trivial for someone to re-do the existing hal implementations for the new traits. And the nice thing is they can technically implement both alongside as well.

I don't think that expecting HALs to implement special traits just for our crate is a scalable solution. If the new traits turn out to be superior to `display-interface` we might as well try to convince other display driver crates to use them instead.

> Once we are reasonably confident in the trait we can release `mipidsi-interface` (or whatever) as instant 1.0.0.

The "or whatever" is definitely the way to go, because `mipidsi-interface` is very misleading, because this will still not have anything to do with the actual MIPI DSI.

---

I didn't have time to look into the new traits in detail, but `set_pixels` being generic over `N` and still using an iterator doesn't seem to be the perfect solution. Why do we need to use iterators if we can just pass slices to the pixel data around? It might monomorphize into the same code in some cases, but with `IntoIterator` we need to unnecessarily rebuffer the pixel data in the `Interface` impls when the pixel data is available as a buffer of bytes in the first place.

File: mipidsi/src/interface/mod.rs
Comment (GrantM11235) [reply mipidsi/src/interface/mod.rs:12]: I agree. Do you have any suggestions?

Comment (almindor) [reply mipidsi/src/interface/mod.rs:12]: Just `Word` ?

**Comment by almindor** [comment]
On the topic of separate crate. If we do not attempt any kind of unification outside of this driver (mipidsi), then I suggest we make this interface private (as in not really needed to be exposed to the user explicitly) and hide it behind the builder.

Users would then just do either `Builder::spi(...)...` or `Builder::new().with_spi()...` and that's it. Everything else would be sort of hidden.

Unless I'm forgetting something, users don't really need to care about the DI. We'd have to make monomorphised builder due to `release` but that's not such a big problem.

File: mipidsi/src/interface/parallel.rs
Comment (GrantM11235) [reply mipidsi/src/interface/parallel.rs]: That was just copy-pasted from display-interface-parallel-gpio. I updated it to clarify that "all pins" is supposed to mean all data bus pins

File: mipidsi/src/error.rs
Comment (GrantM11235) [reply mipidsi/src/error.rs]: I renamed both variants as well as moving it to the builder module where it is actually used

**Comment by GrantM11235** [comment]
> I didn't have time to look into the new traits in detail, but set_pixels being generic over N and still using an iterator doesn't seem to be the perfect solution.

`[Word; N]` is supposed to represents one pixel[^1], so N will be between 1 and 3. When sending rgb565 pixels over an 8 bit interface, we will always send data as pairs of bytes, so we will use an iterator of `[u8; 2]`. I suspect (but haven't tested) that this is more efficient compared to just an iterator of bytes because the compiler doesn't need to worry about running out of buffer space mid-pixel.

> Why do we need to use iterators if we can just pass slices to the pixel data around?

99% of the time we can't just use slices of pixels. For starters, embedded-graphics only uses iterators. Even if we did have a slice of pixels, we would still need to convert rgb565 to big-endian or convert rgb666 to left-aligned. Even if the pixels were already in the right format, they might still need to be copied to a dma capable buffer

> but with IntoIterator we need to unnecessarily rebuffer the pixel data in the Interface impls

Using an iterator at the interface boundary instead of a slice means that we *don't* need to unnecessarily buffer the data when using an unbuffered interface like `ParallelInterface`

[^1]: If we add support for rgb444, it will probably involve sending pairs of pixels packed into `[u8; 3]`, but I'm not sure exactly how or if that would work at this point.

**Comment by GrantM11235** [comment]
One of my design decisions when making this trait was to only support mipi dcs style interfaces, not all "displays" that have some concept of "commands" and "data". The reason is that other display interfaces work differently and have different usage patterns. For example, SSD1306 oled command parameters are sent as "commands" but in mipi dcs they are "data".

That means the only other crates that can *use* this trait are the other crates that support the same displays as us, such as `ili9341`. They are welcome to use the trait, but they are also welcome to deprecate themselves and just use our crate instead :shrug: 

---

We can't make the trait private/sealed because we want to allow high performance, hardware specific impls like stm32's fsmc, rpi's pio, or esp32's i80. These impls don't actually need to be in the HALs, they can be in third party crates as long as the HALs support the basic operations.

We could still add convenience constructors like `Builder::new_spi(...)` but I'll leave that for another day.

File: mipidsi/src/interface/mod.rs
Comment (GrantM11235) [reply mipidsi/src/interface/mod.rs:12]: `Word` seems a bit ambiguous, but ambiguous is better than misleading

**Review by rfuest — CHANGES_REQUESTED** [review]
While I might not like all the details of the implementation I won't oppose merging it. Let's see how it works out in practice. And it's definitely a nice improvement over `display_interface` :+1: .

But one thing I would like to bring up is the mutable slice that is used as a buffer in `SpiInterface`. There might be use cases where this would be harder to work with, because a `Display` that uses `SpiInterface` couldn't be stored in the same struct as the SPI buffer. What is the reason you decided to use this approach instead of making the buffer a const generic inside the `SpiInterface`?

File: mipidsi/src/interface/parallel.rs
Comment (GrantM11235) [reply mipidsi/src/interface/parallel.rs:174]: I don't believe that is a problem in this case. Unlike CS, WR only cares about the rising *edge*. For each word we set it low, which may or may not be a falling edge, then we set it high, which is definitely a rising edge.

Comment (rfuest) [reply mipidsi/src/interface/parallel.rs:174]: OK, that's a valid point. I prefer the pins to be in a known state, but because we don't control CS driving WR high in `new` could even cause an unwanted rising edge on WR. I think we should revisit this later and add optional support for driving CS. If we decide to add read support the correct initial state will become more relevant because the display controller uses the falling edge of RD to put data on the bus.

Comment (GrantM11235) [reply mipidsi/src/interface/parallel.rs:174]: That reminds me, leaving CS and RD floating is a pretty big footgun. Maybe we should require both of them by default. I'll see if I can make the ergonomics work

**Comment by GrantM11235** [comment]
> But one thing I would like to bring up is the mutable slice that is used as a buffer in SpiInterface. There might be use cases where this would be harder to work with, because a Display that uses SpiInterface couldn't be stored in the same struct as the SPI buffer. What is the reason you decided to use this approach instead of making the buffer a const generic inside the SpiInterface?

Some chips can only dma from certain memory regions. If the user is able to create the buffer in one of those regions, then their hal's spi impl won't need to copy the data into another buffer.

I changed it to be generic over `AsMut<[u8]>`, so now you can use `&mut [u8]`, `[u8; N]`, or even `Box<[u8]>`, but I'm not sure if it is worth the extra confusion and I'm worried it could encourage people to use a large array and blow up their stack.

Initializing a display with an inline array buffer involves creating the array, moving it into the interface, then moving that interface into the display driver. Each of those moves involves allocating space for the whole array on the stack and copying the whole array. LLVM is notoriously bad at optimizing these copies away. Moving the display driver after that will have the same problem.

For practical purposes it's probably easier to just use `static-cell` to get a `&'static mut [u8]`. Then the interface/display is also `'static` and can be moved around freely

**Comment by almindor** [comment]
> > But one thing I would like to bring up is the mutable slice that is used as a buffer in SpiInterface. There might be use cases where this would be harder to work with, because a Display that uses SpiInterface couldn't be stored in the same struct as the SPI buffer. What is the reason you decided to use this approach instead of making the buffer a const generic inside the SpiInterface?
> 
> Some chips can only dma from certain memory regions. If the user is able to create the buffer in one of those regions, then their hal's spi impl won't need to copy the data into another buffer.
> 
> I changed it to be generic over `AsMut<[u8]>`, so now you can use `&mut [u8]`, `[u8; N]`, or even `Box<[u8]>`, but I'm not sure if it is worth the extra confusion and I'm worried it could encourage people to use a large array and blow up their stack.
> 
> Initializing a display with an inline array buffer involves creating the array, moving it into the interface, then moving that interface into the display driver. Each of those moves involves allocating space for the whole array on the stack and copying the whole array. LLVM is notoriously bad at optimizing these copies away. Moving the display driver after that will have the same problem.
> 
> For practical purposes it's probably easier to just use `static-cell` to get a `&'static mut [u8]`. Then the interface/display is also `'static` and can be moved around freely

Correct, when it comes to CS there are essentially two scenarios. Either the CS is handled by the SPI/HAL directly or by us. In general we should favour the SPI/HAL to handle it because at least some MCUs have HW control of CS built-in to the "fast SPI" implementation. Note that many HALs have this implemented "wrong" and do not automatically switch CS and other SPI settings (such as mode, speed etc.) properly when switching between `SpiDevice` instances, but that's a problem out of scope for this driver.

If we control CS we need to ensure it's reset properly at the end of a transaction of course. If we do not, the best we can do is just document that it's the upstream's responsibility to not let CS hang.

DMA buffers are a must, the speed increase when using them is quite large.

**Comment by GrantM11235** [comment]
I'm confused, the message you are replying to isn't about CS

**Comment by almindor** [comment]
> I'm confused, the message you are replying to isn't about CS

Up, misclicked. I read parts in email and then came here and clicked the wrong thing. It was meant mostly as a reaction to the CS discussion, but also l inked to the DMA buffer requirement.

File: mipidsi/src/interface/parallel.rs
Comment (GrantM11235) [reply mipidsi/src/interface/parallel.rs:174]: That's going to require more effort, let's leave it for later

**Review by almindor — APPROVED** [review]
I think I'm happy with this change atm. @rfuest I'll let you review one more time since your changes were done, but I'm inclined to go ahead with this and merge soon to get the ball rolling.

**Comment by rfuest** [comment]
> @rfuest I'll let you review one more time since your changes were done, but I'm inclined to go ahead with this and merge soon to get the ball rolling.

The PR is in already in a very good state, but there are still 2 unresolved conversations that should be addressed before this can be merged. The one about the duplicate name isn't too important, but I think that the one about the `flush` method needs to be discussed.

File: mipidsi/src/interface/mod.rs
Comment (GrantM11235) [reply mipidsi/src/interface/mod.rs:49]: I'm leaning towards removing it for now too, but adding it back later is technically a breaking change.

It wouldn't break any *implementers* of the trait, but all *users* of the trait will need to add flushes. This isn't a problem if `mipidsi` is the only user because we can just add the flushes when we update the trait, but it might cause problems for anyone who is using the trait directly.

I think it's probably unlikely to cause a problem in practice, so we could just pretend that it is non-breaking.

Comment (rfuest) [reply mipidsi/src/interface/mod.rs:49]: That's a good point, I hadn't considered that other crates might be *users* of the trait at this point in time. My suggestion would to ignore the possible breaking changes at the moment and deal with them later, if necessary.

The current solution, where the `Interface` trait is defined in the `mipidsi` crate isn't a good long term solution anyway in this regard. A breaking change in any part of `mipidsi` will make all external `Interface` impls incompatible, even if the change wasn't related to the `Interface` trait.

**Review by rfuest — APPROVED** [review]

**Pull request merged**
